subsys/net/lib/sockets/sockets_packet.c

```diff
@@ -0,0 +1,339 @@
+				int status,
+				void *user_data)
+{
+	NET_DBG("ctx=%p, pkt=%p, st=%d, user_data=%p", ctx, pkt, status,
+		user_data);
+
+	/* if pkt is NULL, EOF */
+	if (!pkt) {
+		struct net_pkt *last_pkt = k_fifo_peek_tail(&ctx->recv_q);
+
+		if (!last_pkt) {
+			/* If there're no packets in the queue, recv() may
+			 * be blocked waiting on it to become non-empty,
+			 * so cancel that wait.
+			 */
+			sock_set_eof(ctx);
+			k_fifo_cancel_wait(&ctx->recv_q);
+			NET_DBG("Marked socket %p as peer-closed", ctx);
+		} else {
+			net_pkt_set_eof(last_pkt, true);
+			NET_DBG("Set EOF flag on pkt %p", ctx);
+		}
+
+		return;
+	}
+
+	/* Normal packet */
+	net_pkt_set_eof(pkt, false);
+
+	k_fifo_put(&ctx->recv_q, pkt);
+}
+
+static int zpacket_bind_ctx(struct net_context *ctx,
+			    const struct sockaddr *addr,
+			    socklen_t addrlen)
+{
+	int ret = 0;
+
+	ret = net_context_bind(ctx, addr, addrlen);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	/* For packet socket, we expect to receive packets after call
+	 * to bind().
+	 */
+	ret = net_context_recv(ctx, zpacket_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (ret < 0) {
+		errno = -ret;
+		return -1;
+	}
+
+	return 0;
+}
+
+ssize_t zpacket_sendto_ctx(struct net_context *ctx, const void *buf, size_t len,
+			   int flags, const struct sockaddr *dest_addr,
+			   socklen_t addrlen)
+{
+	s32_t timeout = K_FOREVER;
+	int status;
+
+	if (!dest_addr) {
+		errno = EDESTADDRREQ;
+		return -1;
+	}
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	/* Register the callback before sending in order to receive the response
+	 * from the peer.
+	 */
+
+	status = net_context_recv(ctx, zpacket_received_cb, K_NO_WAIT,
+				  ctx->user_data);
+	if (status < 0) {
+		errno = -status;
+		return -1;
+	}
+
+	status = net_context_sendto_new(ctx, buf, len, dest_addr,
+					addrlen, NULL, timeout, NULL,
+					ctx->user_data);
+	if (status < 0) {
+		errno = -status;
+		return -1;
+	}
+
+	return status;
+}
+
+ssize_t zpacket_recvfrom_ctx(struct net_context *ctx, void *buf, size_t max_len,
+			     int flags, struct sockaddr *src_addr,
+			     socklen_t *addrlen)
+{
+	size_t recv_len = 0;
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (flags & ZSOCK_MSG_PEEK) {
+		int res;
+
+		res = _k_fifo_wait_non_empty(&ctx->recv_q, timeout);
+		/* EAGAIN when timeout expired, EINTR when cancelled */
+		if (res && res != -EAGAIN && res != -EINTR) {
+			errno = -res;
+			return -1;
+		}
+
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+	} else {
+		pkt = k_fifo_get(&ctx->recv_q, timeout);
+	}
+
+	if (!pkt) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* We do not handle any headers here,
+	 * just pass the whole packet to caller.
+	 */
+	recv_len = net_pkt_get_len(pkt);
+	if (recv_len > max_len) {
+		recv_len = max_len;
+	}
+
+	net_frag_linearize(buf, recv_len, pkt, 0, recv_len);
```

Comment:
Use the new API directly, net_pkt_read_new(). Check what I have done in #12989 (see the 2 last commits on udp, tcp for socket). 
(The only trick is to have a backup of the cursor in case ZSOCK_MSG_PEEK)

Reply:
According to your suggestion, if packet has 200 bytes, and caller tries to read (recvfrom) 100 bytes, means **error**. I am bit confused.

Reply:
why? you did already the right check above line 216. You have to keep that of course.

Reply:
Ok.

Reply:
Done, please take a look again.